ecc/bls12-377/shplonk/shplonk.go

```diff
@@ -0,0 +1,462 @@
+	ErrVerifyOpeningProof    = errors.New("can't verify batch opening proof")
+)
+
+// OpeningProof KZG proof for opening (fᵢ)_{i} at a different points (xᵢ)_{i}.
+//
+// implements io.ReaderFrom and io.WriterTo
+type OpeningProof struct {
+
+	// W = ∑ᵢ γⁱZ_{T\Sᵢ}(f_i(X)-r)/Z_{T} where Z_{T} is the vanishing polynomial on the (Sᵢ)_{i}
+	// and r interpolates fᵢ(Sᵢ) on (Sᵢ)
+	W bls12377.G1Affine
+
+	// L(X)/(X-z) where L(X)=∑ᵢγⁱZ_{T\xᵢ}(f_i(X)-rᵢ) - Z_{T}W(X)
+	WPrime bls12377.G1Affine
+
+	// ClaimedValues[i] are the values of fᵢ on Sᵢ
+	ClaimedValues [][]fr.Element
+}
+
+// BatchOpen opens the list of polynomials on points, where the i-th polynomials is opend at points[i].
+func BatchOpen(polynomials [][]fr.Element, digests []kzg.Digest, points [][]fr.Element, hf hash.Hash, pk kzg.ProvingKey, dataTranscript ...[]byte) (OpeningProof, error) {
+
+	var res OpeningProof
+
+	nbInstances := len(polynomials)
+	if len(polynomials) != len(points) {
+		return res, ErrInvalidNumberOfPoints
+	}
+
+	// transcript
+	fs := fiatshamir.NewTranscript(hf, "gamma", "z")
+
+	// derive γ
+	gamma, err := deriveChallenge("gamma", points, digests, fs, dataTranscript...)
+	if err != nil {
+		return res, err
+	}
+
+	// compute the size of the linear combination
+	maxSizePolys := len(polynomials[0])
+	for i := 1; i < len(polynomials); i++ {
+		if maxSizePolys < len(polynomials[i]) {
+			maxSizePolys = len(polynomials[i])
+		}
+	}
+	nbPoints := 0
+	sizeSi := make([]int, len(points))
+	for i := 0; i < nbInstances; i++ {
+		nbPoints += len(points[i])
+		sizeSi[i] = len(points[i])
+	}
+	totalSize := maxSizePolys + nbPoints // upper bound of the size of f := ∑ᵢ γⁱZ_{T\Sᵢ}(f_i(X)-r)
+
+	bufMaxSizePolynomials := make([]fr.Element, maxSizePolys)
+	bufTotalSize := make([]fr.Element, totalSize)
+	f := make([]fr.Element, totalSize) // cf https://eprint.iacr.org/2020/081.pdf page 11 for notation
+	res.ClaimedValues = make([][]fr.Element, nbInstances)
+	for i := 0; i < nbInstances; i++ {
+		res.ClaimedValues[i] = make([]fr.Element, len(points[i]))
+	}
+	var accGamma fr.Element
+	accGamma.SetOne()
+
+	ztMinusSi := make([][]fr.Element, nbInstances)
+	ri := make([][]fr.Element, nbInstances)
+	for i := 0; i < nbInstances; i++ {
+
+		for j := 0; j < len(points[i]); j++ {
+			res.ClaimedValues[i][j] = eval(polynomials[i], points[i][j])
+		}
+
+		ztMinusSi[i] = buildZtMinusSi(points, i)
+
+		copy(bufMaxSizePolynomials, polynomials[i])
+		ri[i] = interpolate(points[i], res.ClaimedValues[i])
+		sub(bufMaxSizePolynomials, ri[i])
+
+		bufTotalSize = mul(bufMaxSizePolynomials, ztMinusSi[i], bufTotalSize)
+		bufTotalSize = mulByConstant(bufTotalSize, accGamma)
+		for j := 0; j < len(bufTotalSize); j++ {
+			f[j].Add(&f[j], &bufTotalSize[j])
+		}
+
+		accGamma.Mul(&accGamma, &gamma)
+		setZero(bufMaxSizePolynomials)
+	}
+
+	zt := buildVanishingPoly(flatten(points))
+	w := div(f, zt) // cf https://eprint.iacr.org/2020/081.pdf page 11 for notation page 11 for notation
```

Comment:
on page 11 of the paper they name the committed quotient W, and the vanishing polynomial zt

Reply:
Yup, the committed quotient is `W` and the vanishing polynomial is `zt`. But the polynomial `f/zt` is not implicitly named on page 11 (they only use `f/zt` notation), but it is on page 9 (as `h = f/zt`).

As you use `w` (or `h`) later in step 4 (p 11), then to follow paper imo it makes more sense to rename `w` to `h`. But it is a very insignificant difference, it can stay as is.